File: TP02/questao6/questao6.c
#include <stdio.h>
#include <string.h>
#include <stdlib.h>
#include <math.h>
#include <stdbool.h>
#include <time.h>

//Struct que representa a classe Jogador

typedef struct Jogador
{
    // Atributos
     int id;
     char* nome;
     int altura;
     int peso;
     char* universidade;
     int anoNascimento;
     char* cidadeNascimento;
     char* estadoNascimento;

}Jogador;


/*Procedimento que imprime todo conteudo de uma Struct*/

void imprimir(Jogador* jogador){
    if(jogador == NULL){
        printf("Jogador inválido");
    }
    else{
    printf("[%d ## %s ## %d ## %d ## %d ## %s ## %s ## %s]\n",
           jogador->id,
           jogador->nome,
           jogador->altura,
           jogador->peso,
           jogador->anoNascimento,
           jogador->universidade,
           jogador->cidadeNascimento,   
           jogador->estadoNascimento);
    }
}

//Simulando construtores

void constructor(Jogador* jogador, int id, char* nome, int altura, int peso, char* universidade, int anoNascimento, char* cidadeNascimento, char* estadoNascimento){
    
    if(jogador != NULL){
        jogador->id = id;
        jogador->nome = nome;
        jogador->altura = altura;
        jogador->peso = peso;
        jogador->universidade = universidade;
        jogador->anoNascimento = anoNascimento;
        jogador->cidadeNascimento = cidadeNascimento;
        jogador->estadoNascimento = estadoNascimento;
    }
}

void emptyConstructor(Jogador* jogador){
    constructor(jogador,-1,"",0,0,"",0,"","");
}



/*Procedimento que extrai as colunas da linha, substituindo valores vazios por "nao informado"
 * @params: Vetor de apontadores que vai receber as Strings separadas, Apontador para cadeia de caracteres contendo a linha
 */

void tratamentoLinha(char** colunas, char* linha){
    colunas[0] = strtok(linha, ",");
    char* prevTokenEnd = colunas[0] + strlen(colunas[0]);

    for (int i = 1; i < 8; i++){
        if (*(prevTokenEnd + 1) == ',') {
            colunas[i] = "nao informado";
            prevTokenEnd++;
        } else {
            colunas[i] = strtok(NULL, ",");
            if (colunas[i]) {
                prevTokenEnd = colunas[i] + strlen(colunas[i]);
            } else {
                colunas[i] = "nao informado";
            }
        }
    }
}

/*Procedimento que instancia Struct com as informações proveninentes de uma linha (representada por um ID) do arquivo
 @params: Apontador para struct, nome do arquivo a ser lido, id do jogador a ser criado
*/

Jogador** lerParaStruct(char* nomeArquivo, int* lista_ids, int tamanho_lista){
    FILE* file = fopen(nomeArquivo,"r");
    if(file == NULL){
        printf("Erro ao abrir arquivo\n");
        exit(0);
    }
    
    //Apotandor para vetor de strings contendo as linhas
    
    char** linhas = (char**)malloc(sizeof(char*) * 4000);

    //Alocando espaço para cada linha do vetor

    for(int i = 0; i < 4000; i++){
        linhas[i] = (char*)malloc(sizeof(char) * 256);
    }

    //Inicializando cada posicao do vetor com as linhas do arquivo

    int contador = 0;

    while(fscanf(file, " %255[^\n]", linhas[contador]) == 1){
        //printf("Linha lida: %s\n", linhas[contador]);
        contador++;
    }


    //Crinando o vetor das colunas e alocando espaço para cada linha

    char** colunas = (char**)malloc(sizeof(char*) * 8);

    for(int i = 0; i < 8; i++){
        colunas[i] = (char*)malloc(sizeof(char) * 75);
    }
    
    //Vetor para armazenar todos os jogadores

    Jogador** todos_jogadores = (Jogador**)malloc(sizeof(Jogador*) * contador);

    //Extraindo as colunas 
    
    for(int i = 1; i < contador; i++){
        Jogador* jogador = (Jogador*)malloc(sizeof(Jogador));
        tratamentoLinha(colunas,linhas[i]);
        int id = atoi(colunas[0]);
        int altura = atoi(colunas[2]);
        int peso = atoi(colunas[3]);
        int ano = atoi(colunas[5]);

        constructor(jogador, id, colunas[1], altura, peso, colunas[4], ano, colunas[6], colunas[7]);
    
        todos_jogadores[i-1] = jogador;
    }

    //Vetor para armazenar os ids corretos

    Jogador** jogadores_corretos = (Jogador**)malloc(sizeof(Jogador*) * tamanho_lista);

    //Criando vetor com os ids corretos

    int k = 0;

    for(int i = 0; i < contador-1; i++){
        for(int j = 0; j < tamanho_lista; j++){
            if(todos_jogadores[i]->id == lista_ids[j]){
                jogadores_corretos[k] = todos_jogadores[i];
                k++;
            }
        }
    }

    fclose(file);
   
    return jogadores_corretos;
}

void swap(Jogador* jogador1, Jogador* jogador2){
    Jogador temp = *(jogador1);
    *(jogador1) = *(jogador2); 
    *(jogador2) = temp;
}


void selecao_nome_recursiva(Jogador** Jogadores, int size, int i){
    if(i == size - 1){
        return;
    }
    int menor = i;
    for(int j = (i + 1); j < size; j++){
        if(Jogadores[j] != NULL && Jogadores[menor] != NULL){
            char* nome_j = Jogadores[j]->nome;
            char* nome_menor = Jogadores[menor]->nome;
            int comp = strcmp(nome_j,nome_menor);
            if(comp < 0){
                menor = j;
            }
        }
    }
    swap(Jogadores[menor],Jogadores[i]);
    selecao_nome_recursiva(Jogadores,size,i + 1);
}
  




int main(){


    //SEÇÃO QUE CAPUTRA OS IDS

    Jogador* jogador = (Jogador*)malloc(sizeof(Jogador));

    int* lista_ids = (int*)malloc(sizeof(int) * 100);

    char* ids = (char*)malloc(sizeof(char) * 10); 

    scanf("%s",ids);

    int i = 0;

    while(strcmp(ids,"FIM") != 0){
        lista_ids[i] = atoi(ids);
        i++;
        scanf("%s",ids);
        getchar();
    }
    


    Jogador** jogadores_para_ordenar = lerParaStruct("/tmp/players.csv",lista_ids,i);

    selecao_nome_recursiva(jogadores_para_ordenar,i,0);

    for(int j = 0; j < i - 1; j++){
        //printf("%d",j);
        imprimir(jogadores_para_ordenar[j]);
    }

    return 0;
}

File: TP03/Questao6/Questao6.java
import java.io.BufferedReader;
import java.io.EOFException;
import java.io.FileReader;
import java.util.ArrayList;
import java.util.List;
import java.util.Scanner;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

class Resultado {
    public Pilha jogadoresAtualizados;
    public Jogador jogadoresRemovidos;

    public Resultado() {
        this.jogadoresAtualizados = new Pilha();
        this.jogadoresRemovidos = new Jogador();
    }

    @Override
    public String toString() {
        StringBuilder sb = new StringBuilder();

        sb.append("Jogadores Atualizados:\n");
        sb.append(jogadoresAtualizados.toString()); // Supondo que a classe ArrayList<Jogador> também tenha um método toString().
        sb.append("\nJogadores Removidos:\n");
        sb.append(jogadoresRemovidos.toString());

        return sb.toString();
    }

}

class Celula {
    public Celula prox;
    public Jogador jogador;

    public Celula() {
        this.jogador = null;
        this.prox = null;
    }

    public Celula(Jogador jogador) {
        this.jogador = jogador;
        this.prox = null;
    }
}

class Pilha { // ArrayList<Jogador> para o tipo Jogador
    // Atributos
    private Celula topo;
    private int contador;

    // Metodos

    public  Pilha() {
        this.topo = null;
    }

    public int size() {
        return this.contador;
    }

    public void push(Jogador jogador) {
        Celula temp = new Celula(jogador);
        temp.prox = topo;
        topo = temp;
        contador++;
    }


    public Jogador pop() {
        if(topo == null) {
            throw new Error("LISTA VAZIA");
        }
        else {
            Jogador removido = topo.jogador;
            topo = topo.prox;
            contador--;
            return removido;
        }
    }

    public Jogador get(int i) {
        Celula j = topo;
        for(int k = 0; k < i; k++, j = j.prox);
        Jogador jogador = j.jogador;
        return jogador;
    }

}

class Jogador {

    // Atributos
    private int id;
    private String nome;
    private int altura;
    private int peso;
    private String universidade;
    private int anoNascimento;
    private String cidadeNascimento;
    private String estadoNascimento;

    // Construtores

    public Jogador() {

    }

    public Jogador(int id, String nome, int altura, int peso, String universidade, int anoNascimento,
            String cidadeNascimento, String estadoNascimento) {

        this.id = id;
        this.nome = nome;
        this.altura = altura;
        this.peso = peso;
        this.universidade = universidade;
        this.anoNascimento = anoNascimento;
        this.cidadeNascimento = cidadeNascimento;
        this.estadoNascimento = estadoNascimento;
    }

    // Métodos

    public void setId(int id) {
        this.id = id;
    }

    public int getId() {
        return this.id;
    }

    public void setNome(String nome) {
        this.nome = nome;
    }

    public String getNome() {
        return this.nome;
    }

    public void setAltura(int altura) {
        this.altura = altura;
    }

    public int getAltura() {
        return this.altura;
    }

    public void setUniversidade(String universidade) {
        this.universidade = universidade;
    }

    public String getUniversidade() {
        return this.universidade;
    }

    public void setAnoNascimento(int anoNascimento) {
        this.anoNascimento = anoNascimento;
    }

    public int getAnoNascimento() {
        return this.anoNascimento;
    }

    public void setCidadeNascimento(String cidadeNascimento) {
        this.cidadeNascimento = cidadeNascimento;
    }Jogador jogador;

    public String getCidadeNascimento() {
        return this.cidadeNascimento;
    }

    public void setEstadoNascimento(String estadoNascimento) {
        this.estadoNascimento = estadoNascimento;
    }

    public String getEstadoNascimento() {
        return this.estadoNascimento;
    }

    protected Jogador clone() throws CloneNotSupportedException {
        Jogador novo = new Jogador();
        novo.id = this.id;
        novo.nome = this.nome;
        novo.altura = this.altura;
        novo.peso = this.peso;
        novo.universidade = this.universidade;
        novo.anoNascimento = this.anoNascimento;
        novo.cidadeNascimento = this.cidadeNascimento;
        novo.estadoNascimento = this.estadoNascimento;
        return novo;
    }
    /*
     * Método que extrai as colunas da linha, substituindo valores vazios por
     * "nao informado"
     * 
     * @params: String linha
     * 
     * @return: Array de Strings colunas
     */

    public static String[] tratamentoString(String str) {
        List<String> colunas = new ArrayList<>();
        Matcher matcher = Pattern.compile("(\"[^\"]*\"|[^,]*),?").matcher(str);

        while (matcher.find()) {
            String coluna = matcher.group(1);
            if (coluna != null && !coluna.isEmpty()) {
                // Remove as aspas se estiver entre aspas
                coluna = coluna.startsWith("\"") && coluna.endsWith("\"") ? coluna.substring(1, coluna.length() - 1)
                        : coluna;
            } else {
                coluna = "nao informado";
            }
            colunas.add(coluna);
        }

        // Garante que temos exatamente 8 colunas
        while (colunas.size() < 8) {
            colunas.add("nao informado");
        }

        return colunas.toArray(new String[0]);
    }

    /*
     * Método que itera pelo arquivo players.csv, criando um vetor com todos os
     * objetos requisitado
     * 
     * @param array list de inteiros contendo os ids requisitados, String nome do
     * arquivo,
     * 
     * @return: array list de jogadores contendo os jogadores referentes aos Ids
     * requisitados
     */

    public static ArrayList<Jogador> criarVetor(String nomeArquivo) throws Exception {

        // Le todo o arquivo e concatena todo o conteúdo em uma String

        FileReader file = new FileReader(nomeArquivo);
        BufferedReader buffer = new BufferedReader(file);
        String json = "";
        String line = buffer.readLine();
        while (line != null) {
            json = json + line + "\n";
            line = buffer.readLine();
        }

        buffer.close();
        file.close();

        // Converte o conteúdo do arquivo para um array de linhas
        String[] linhas = json.split("\n");

        ArrayList<Jogador> Jogadores = new ArrayList<>();

        String[] colunas = new String[8];

        // Loop que armazena todos os jogadores em uma ArrayList<Jogador>

        for (int i = 1; i < linhas.length; i++) {
            // System.out.println(linhas[i]);

            colunas = tratamentoString(linhas[i]);

            int id = Integer.parseInt(colunas[0]);
            if (id == 223) {
                Jogadores.add(null);
                continue; // Ignora a entrada se o id for 223
            }
            // Instanciando o objeto
            Jogador objeto = new Jogador(
                    Integer.parseInt(colunas[0]),
                    colunas[1],
                    Integer.parseInt(colunas[2]),
                    Integer.parseInt(colunas[3]),
                    colunas[4],
                    Integer.parseInt(colunas[5]),
                    colunas[6],
                    colunas[7]);

            Jogadores.add(objeto);
            // System.out.println("JOGADOR SENDO PRINTADO: " + objeto.id);
            // imprimir(Jogadores.get(i - 1));
        }

        return Jogadores;

    }
    /*
     * Método que filtra os jogadores por ID, retornando uma ArrayList<Jogador> de jogadores
     * corretos
     */

    public static Pilha filtraId(ArrayList<Integer> ids, ArrayList<Jogador> Jogadores) throws Exception {
        Pilha Jogadores_corretos = new Pilha();

        Jogador[] array = new Jogador[4000]; 

        for(int i = 0; i < Jogadores.size(); i++) {
            array[i] = Jogadores.get(i);
        }


        for (int i = 0; i < ids.size(); i++) {
            int id = ids.get(i);
            for (int j = 0; j < Jogadores.size(); j++) {
                if (array[j] != null) {
                    Jogador jogador = Jogadores.get(j);
                    if (jogador.id == id) {
                        Jogadores_corretos.push(jogador);
                        //imprimir(jogador);
                    }
                }
            }
        }
        return Jogadores_corretos;
    }

    /*
     * Metodo que recebe a entrada e divide ela em: comando E numero
     * 
     * @params String contendo o comando e o numero
     * 
     * @return Array de String contendo respectivamente o comando e o numero
     */

    public static String[] trataString(String comando) {
        // System.out.println(comando);
        String[] comando_tratado = new String[2]; // Aumentar para 3, para acomodar possível posição em comandos I*

        if (comando.startsWith("I")) {
            comando_tratado[0] = comando.substring(0, 1);
            comando_tratado[1] = comando.substring(2);
        }
        else {
            comando_tratado[0] = comando.substring(0, 1);
            comando_tratado[1] = "";
        } 

        for (int i = 0; i < 2; i++) {
            //System.out.println(comando_tratado[i]);
        }

        return comando_tratado;
    }

    /*
     * Metodo que efetivamente realiza as chamads das funcoes na Lista de jogadores
     * com base na String tratada
     */

    public static Resultado callCommand(String[] comando_numero, Pilha jogadores, ArrayList<Jogador> todos_jogadores) throws Exception {
        String comando = comando_numero[0];
        Resultado res = new Resultado();

        switch (comando) {
            case "I":
                int id = Integer.parseInt(comando_numero[1]);
                Jogador jogadorInicio = todos_jogadores.get(id);
                jogadores.push(jogadorInicio);
                //imprimir(p.get(p.size() - 1));
                res.jogadoresRemovidos = null;
                break;

            case "R":
                Jogador removido = jogadores.pop();
                //imprimir(removido);
                res.jogadoresRemovidos = removido;
                break;
            
            default:
                break;
        }

        res.jogadoresAtualizados = jogadores;

        return res;
    }


    /* Método void que imprime um Jogador usado apenas para fins de Debug */

    public static void imprimir(Jogador objeto) {
        if (objeto != null) {
            System.out.println("[" + objeto.id + " ## " + objeto.nome + " ## " + objeto.altura + " ## " + objeto.peso
                    + " ## " + objeto.anoNascimento
                    + " ## " + objeto.universidade + " ## " + objeto.cidadeNascimento + " ## " + objeto.estadoNascimento
                    + " ]");
        } else {
            System.out.println("NULL PEBA");
        }
    }

    public static void imprimir_verde(Pilha jogadores) throws Exception {
        for (int i = jogadores.size() - 1; i >= 0; i--) {
            Jogador jogador = jogadores.get(i);
            System.out.println("[" + ((jogadores.size() - 1) - i) + "] ## " + jogador.nome + " ## " + jogador.altura + " ## " + jogador.peso
                    + " ## " + jogador.anoNascimento
                    + " ## " + jogador.universidade + " ## " + jogador.cidadeNascimento + " ## "
                    + jogador.estadoNascimento + " ##");

        }
    }

}



public class Questao6 extends Jogador {

    static Scanner sc = new Scanner(System.in);


    public static void main(String[] args) throws Exception {

        String numero = sc.nextLine();

        ArrayList<Integer> ids = new ArrayList<>();

        ArrayList<Jogador> todos_Jogadores = new ArrayList<>();

        Pilha Jogadores = new Pilha();

        // Bloco que recebe as entradas dos ids a serem adicionados no vetor

        while (!numero.equals("FIM")) {
            int inteiro = Integer.parseInt(numero);
            ids.add(inteiro);
            numero = sc.nextLine();
        }

        int n = sc.nextInt();

        String[] comandos = new String[1000];

        int contador1 = 0;

        todos_Jogadores = criarVetor("/tmp/players.csv");

        for (int i = 0; i < todos_Jogadores.size(); i++) {
            //imprimir(todos_Jogadores.get(i));
        }

        Jogadores = filtraId(ids, todos_Jogadores);

        for (int i = 0; i < Jogadores.size(); i++) {
            //imprimir(Jogadores.get(i));
        }
        while (sc.hasNextLine()) {
            String comando = sc.nextLine();
            comandos[contador1] = comando;
            // System.out.println(comandos[contador1]);
            contador1++;
        }

        for (int i = 0; i < Jogadores.size(); i++) {
            // imprimir(Jogadores.get(i));
        }
        // loop que chama os comandos

        ArrayList<Jogador> jogadoresRemovidos = new ArrayList<>();

        for (int i = 1; i <= n; i++) {
            // System.out.println(comandos[i]);
            String[] comandos_aux = trataString(comandos[i]);
            Resultado resultado = callCommand(comandos_aux, Jogadores, todos_Jogadores);
            Jogadores = resultado.jogadoresAtualizados;
            if (resultado.jogadoresRemovidos != null) {
                jogadoresRemovidos.add(resultado.jogadoresRemovidos);
            }
        }

        for (int i = 0; i < jogadoresRemovidos.size(); i++) {
            System.out.println("(R) " + jogadoresRemovidos.get(i).getNome());
        }

        imprimir_verde(Jogadores);
    }

}
